### DISCUSSION TITLE
[RFC] Customer collections

### DISCUSSION THREAD

**Discussion by maarcingebala** [saleor/saleor#14418] [RFC]
## Problem

Saleor needs to have a way to group customers into collections, which could be used for multiple purposes such as:
- granting a specific promotion/discount/voucher
- assigning a tax class 
- automatically assign users to collections upon login or first order

## 1. General Assumptions

- Customer collection consists of a name and a collection of assigned customers
- Collections support metadata
- Customer can belong to more than one collection

### Acceptance Criteria

- [ ] Add a model to store the customer collections and the M2M models with relations to User. Populate db should create at least one customer collection. `cleardb` should clear all the customer collections.
- [ ] Add new GraphQL type for the collections
- [ ] Add the queries to retrieve the collections with pagination and data loaders
- [ ] Add mutation to add a collection with relevant errors
- [ ] Add new events for customer collection creation, update, and their triggering code

### Mutations

- All mutations should require `MANAGE_USERS` permission.
- Each should have its own error type class; see `customerCollectionCreate` and `CustomerCollectionCreateError` below (error types for other mutations are skipped in the below schema).
- Each should be placed in a separate file e.g. `customerCollectionCreate` goes to `customer_collection_create.py` etc.
- Types should use dataloaders to load relationship data efficiently.
- Each mutation should trigger a webhook event (names of events are added in the schema descriptions comments). For all new webhooks, we prefer subscription webhooks so each event will need a dedicated event type (see [subscription_types.py](https://github.com/saleor/saleor/blob/main/saleor/graphql/webhook/subscription_types.py) for reference).

```graphql
type Mutation {

  """
  Create a customer collection.
  Requires `MANAGE_USERS` permissions.
  Triggers `CUSTOMER_COLLECTION_CREATED` event.
  """
  customerCollectionCreate(
    input: CustomerCollectionCreateInput!
  ): CustomerCollectionCreate @doc(category: "Users")

  """
  Update a customer collection.
  Requires `MANAGE_USERS` permissions.
  Triggers `CUSTOMER_COLLECTION_UPDATED` event.
  """
  customerCollectionUpdate(
    id: ID!
    input: CustomerCollectionUpdateInput!
  ): CustomerCollectionUpdate @doc(category: "Users")

  """
  Delete a customer collection.
  Requires `MANAGE_USERS` permissions.
  Triggers `CUSTOMER_COLLECTION_DELETED` event.
  """
  CustomerCollectionDelete(
    id: ID!
  ): CustomerCollectionDelete @doc(category: "Users")
}

input CustomerCollectionCreateInput {
  name: String!

  """Slug of the customer collection; if not provided, it's generated from the name."""
  slug: String

  """Description of a collection e.g. it's purpose."""
  description: String

  metadata: [MetadataInput!]
  privateMetadata: [MetadataInput!]
  
  """List of customers to add to this collection."""
  addCustomers: [ID!]
}

input CustomerCollectionUpdateInput {
  name: String
  slug: String
  description: String
  metadata: [MetadataInput!]
  privateMetadata: [MetadataInput!]
  addCustomers: [ID!]
  removeCustomers: [ID!]
}
```

### Types

```graphql
type CustomerCollection {
  id: ID!
  name: String!
  slug: String
  description: String
  createdAt: DateTime!
  updatedAt: DateTime!

  """A list of customers assigned to this collection."""
  customers(
    filter: CustomerFilterInput
    sortBy: UserSortingInput 
    before: String
    after: String
    first: Int
    last: Int
  ): UserCountableConnection @doc(category: "Users")

  privateMetadata: [MetadataItem!]!
  privateMetafield(key: String!): String
  privateMetafields(keys: [String!]): Metadata
  metadata: [MetadataItem!]!
  metafield(key: String!): String
  metafields(keys: [String!]): Metadata
}

type CustomerCollectionCreate @doc(category: "Users") {
  errors: [CustomerCollectionCreateError!]!
  CustomerCollection: CustomerCollection
}

type CustomerCollectionCreateError @doc(category: "Users") {
  field: String
  message: String
  code: CustomerCollectionCreateErrorCode!
  customers: [ID!]
}

enum CustomerCollectionCreateErrorCode @doc(category: "Users") {
  GRAPHQL_ERROR
  INVALID
  NOT_FOUND
  REQUIRED
  UNIQUE
}
```

### Queries

```graphql
"""
List of customer collections. Requires `MANAGE_USERS` permission.
"""
CustomerCollections(
  filter: CustomerCollectionFilterInput
  sortBy: CustomerCollectionSortingInput
  before: String
  after: String
  first: Int
  last: Int
): UserCountableConnection @doc(category: "Users")


input CustomerCollectionFilterInput @doc(category: "CustomerCollections") {
  name: String
}

input CustomerCollectionSortingInput @doc(category: "CustomerCollections") {
  """Specifies the direction in which to sort users."""
  direction: OrderDirection!

  """Sort users by the selected field."""
  field: CustomerCollectionSortField!
}

enum CustomerCollectionSortField @doc(category: "CustomerCollections") {
  """Sort customer collections by name."""
  NAME

  """Sort customer collections by created at."""
  CREATED_AT

  """Sort customer collections by last modified at."""
  LAST_MODIFIED_AT
}


"""
Look up a customer collection by ID. Requires `MANAGE_USERS` permission.
"""
CustomerCollection(
  id: ID!
): CustomerCollection @doc(category: "Users")
```

## 3. Database changes

- Add `CustomerCollection` model:

```
class CustomerCollection:
  name: String
  slug: SlugField, unique
  description: String
  createdAt: DateTime
  updatedAt: DateTime
  customers: ManyToMany(User) // list of customers assigned to this collection
```

## 4. Parked for potential future development
- assigning collections to channels
- disabling/enabling a collection
- mutations to translate customer collection names
- bulk mutations for collections
- allow to assign tax classes to customer collections to override tax for specific collections of people?
- decide what is the expected behaviour of deleting a collection when it has e.g. vouchers connected to it (do they stay assigned to users?)
- dynamic collections (and their naming, because maybe they should be called segments)
- Auto-assign user to customer collection via email domain (e.g. all people who work at a specific company)
- Consider adding `collections` with slugs to `me` and `customer` query. Think of the consequences of exposing the slugs to public and the requirement to make collections private and public in case we have promotions depending on the group names that the end users would be able to overuse.

**Comment by korycins** [comment]
Would be good to mention that the scope of these changes should also include new subscription types for the webhooks (this could not be clear). 
Also missing declaration of a new query to fetch the groups, right? It is probably obvious but worth adding to rfc + required permission.

**Comment by maarcingebala** [comment]
Updated the RFC:
- Added missing `customerGroups` and `customerGroup` queries in the schema examples
- Added information about adding new subscription types in the mutation assumptions section.

**Comment by fowczarek** [comment]
1. Should we think about sorting or filtering customer groups? 
2. Should `cleardb` command removes customer groups?

TODO:
- Populate db should create at least one customer group. 

Ideas to feature development:
- Auto-assign user to customer group via email domain.

**Comment by patrys** [comment]
Nitpick: it's called "collections" now so the API draft needs an update.

**Comment by aniav** [comment]
@patrys changed all the naming to collections now.

@fowczarek added the todo and ideas.
Ad 1. I'd go with minimal sorting and filtering for the first implementation. Added filtering by name and sorting by name and creation date.
Ad 2. `cleardb` does clear all users who are not staff so I'd say for the first implementation it should clear all collections too. If we find it causes problems after initial implementation we will improve.

**Comment by timuric** [comment]
I don't see description of  `CustomerCollectionFilterInput` it would be useful to keep the same filter input as `CustomerFilterInput` https://docs.saleor.io/docs/3.x/api-reference/users/inputs/customer-filter-input

**Comment by timuric** [comment]
Should `me` query and `customer` have property `collections` with slugs? This would be useful for front-end to personalize the experience.

Note: if we expose slug to front-end we should document in the schema that slug is potentially public, to prevent users assigning undesirable names such as `low-value-customer`

**Comment by aniav** [comment]
@timuric added definitions for filters (we have CustomerFilterInput in one place and new `CustomerCollectionFilterInput` in another). Thanks for spotting that!

Added the collections with slugs to parked for future development to not increase the scope of that RFC and keep the changeset minimal 🙌

**Comment by PavelHonza** [comment]
Is here some progress?
